src/terminal/adapter/adaptDispatch.cpp

```diff
@@ -1375,87 +1360,69 @@ bool AdaptDispatch::ScrollDown(const VTInt uiDistance)
-bool AdaptDispatch::_GetParserMode(const StateMachine::Mode mode) const
-{
-    return _api.GetStateMachine().GetParserMode(mode);
 }
 
 // Routine Description:
-// - Sets the various StateMachine parser modes.
+// - Set the alternate screen buffer mode. In virtual terminals, there exists
+//     both a "main" screen buffer and an alternate. This mode is used to switch
+//     between the two.
 // Arguments:
-// - mode - the parser mode to change.
-// - enable - set to true to enable the mode, false to disable it.
+// - enable - true selects the alternate buffer, false returns to the main buffer.
 // Return Value:
 // - <none>
-void AdaptDispatch::_SetParserMode(const StateMachine::Mode mode, const bool enable)
+void AdaptDispatch::_SetAlternateScreenBufferMode(const bool enable)
 {
-    _api.GetStateMachine().SetParserMode(mode, enable);
+    if (enable)
+    {
+        CursorSaveState();
+        _api.UseAlternateScreenBuffer();
+        _usingAltBuffer = true;
+    }
+    else
+    {
+        _api.UseMainScreenBuffer();
+        _usingAltBuffer = false;
+        CursorRestoreState();
+    }
 }
 
 // Routine Description:
-// - Sets the various terminal input modes.
-// Arguments:
-// - mode - the input mode to change.
-// - enable - set to true to enable the mode, false to disable it.
+// - Determines whether we need to pass through input mode requests.
+//   If we're a conpty, AND WE'RE IN VT INPUT MODE, always pass input mode requests
+//   The VT Input mode check is to work around ssh.exe v7.7, which uses VT
+//   output, but not Input.
+//   The original comment said, "Once the conpty supports these types of input,
+//   this check can be removed. See GH#4911". Unfortunately, time has shown
+//   us that SSH 7.7 _also_ requests mouse input and that can have a user interface
+//   impact on the actual connected terminal. We can't remove this check,
+//   because SSH <=7.7 is out in the wild on all versions of Windows <=2004.
 // Return Value:
-// - true if successful. false otherwise.
-bool AdaptDispatch::_SetInputMode(const TerminalInput::Mode mode, const bool enable)
+// - True if we should pass through. False otherwise.
+bool AdaptDispatch::_PassThroughInputModes()
 {
-    _terminalInput.SetInputMode(mode, enable);
-
-    // If we're a conpty, AND WE'RE IN VT INPUT MODE, always pass input mode requests
-    // The VT Input mode check is to work around ssh.exe v7.7, which uses VT
-    // output, but not Input.
-    // The original comment said, "Once the conpty supports these types of input,
-    // this check can be removed. See GH#4911". Unfortunately, time has shown
-    // us that SSH 7.7 _also_ requests mouse input and that can have a user interface
-    // impact on the actual connected terminal. We can't remove this check,
-    // because SSH <=7.7 is out in the wild on all versions of Windows <=2004.
-
-    return !_api.IsConsolePty() || !_api.IsVtInputEnabled();
+    return _api.IsConsolePty() && _api.IsVtInputEnabled();
```

Comment:
I wanted to get rid of `_SetInputMode`, and just call `_terminalInput.SetInputMode` directly, because in many cases we don't care about the passthrough conditions, and these two api calls are wasted. But I've kept them in this helper method, so we can easily use this when we do actually care about passthrough, and also so we don't lose the comment above.